src/core/namespace.c

```diff
@@ -287,10 +288,18 @@ static int apply_mount(
 
                 /* First, get rid of everything that is below if there
                  * is anything... Then, overmount it with an
-                 * inaccessible directory. */
+                 * inaccessible path. */
                 umount_recursive(m->path, 0);
 
-                what = "/run/systemd/inaccessible";
+                // mount() always follow symlinks, so we follow symlinks during the stat() as well
+                r = stat(m->path, &target);
+                if (r != 0)
+                        return r;
+                if (!S_ISDIR(target.st_mode))
+                        what = "/dev/null";
+                else {
+                        what = "/run/systemd/inaccessible";
+                }
                 break;
```

Comment:
> Also for the systemd-nspawn extension how should we proceed? DeviceAllow= does not support major/minor passing. How can we introduce this capability?

Hmm, this should work:

```c
DeviceAllow=/run/inaccessible/blk rwm
DeviceAllow=/run/inaccessible/chr rwm
```

Of course, only after the other work was done, so that those nodes even exist...

Other than that, I think as a special magic we could permit a syntax of "DeviceAllow=c0:0 rwm" and "DeviceAllow=b0:0 rwm" or so, the same way as udev uses it for places such as /run/udev/data/*.